[GLIB] Unify code for accessing default `cairo_font_options`

obyknovenius opened the pull request:
#### 9ccc5caacd23526def9cf7111283b5233be038e1
<pre>
[GLIB] Unify code for accessing default `cairo_font_options`
<a href="https://bugs.webkit.org/show_bug.cgi?id=252057">https://bugs.webkit.org/show_bug.cgi?id=252057</a>

Reviewed by Carlos Garcia Campos.

Currently, there are two different implementations of
`getDefaultCairoFontOptions()` on WPE and GTK.

WPE version simply returns a static value created by
`cairo_font_options_create()`, whereas GTK one is creating a singleton
class that stores and updates an instance variable to keep it in sync
with `GtkSettings`.

This patch makes `getDefaultCairoFontOptions()` simply return a static
value and adds `setDefaultCairoFontOptions()` function. GTK port use it
to set a new value when `GtkSettingsManagerProxy` applies new settings.

* Source/WebCore/platform/graphics/cairo/CairoUtilities.cpp:
(WebCore::getDefaultCairoFontOptions):
(WebCore::setDefaultCairoFontOptions):
* Source/WebCore/platform/graphics/cairo/CairoUtilities.h:
* Source/WebCore/platform/graphics/gtk/GdkCairoUtilities.cpp:
(WebCore::SystemFontOptions::singleton): Deleted.
(WebCore::SystemFontOptions::SystemFontOptions): Deleted.
(WebCore::SystemFontOptions::fontOptions const): Deleted.
(WebCore::SystemFontOptions::updateFontOptions): Deleted.
(WebCore::getDefaultCairoFontOptions): Deleted.
* Source/WebKit/WebProcess/gtk/GtkSettingsManagerProxy.cpp:
(WebKit::GtkSettingsManagerProxy::applySettings):
(WebKit::GtkSettingsManagerProxy::applyFontOptions):
* Source/WebKit/WebProcess/gtk/GtkSettingsManagerProxy.h:

Canonical link: <a href="https://commits.webkit.org/260301@main">https://commits.webkit.org/260301@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/08a0452d7dc884d241b5efdd2669e922a5d5f81a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107684 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16736 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40579 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116831 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116225 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111574 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18158 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8054 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99858 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113437 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13703 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96906 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41387 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95619 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28544 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/83223 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9717 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29893 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10395 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6778 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15871 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49478 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11955 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3866 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/cairo/CairoUtilities.cpp:
```diff
@@ -49,13 +49,17 @@
+    return s_defaultCairoFontOptions.get();
+}
+
+void setDefaultCairoFontOptions(cairo_font_options_t* options)
```
This should receive a `CairoUniquePtr<cairo_font_options_t>&&` to make it clear we are transferring the ownership.

carlosgcampos commented on Source/WebCore/platform/graphics/cairo/CairoUtilities.cpp:
```diff
@@ -49,13 +49,17 @@
 namespace WebCore {
 
-#if USE(CAIRO) && !PLATFORM(GTK)
+static auto s_defaultCairoFontOptions = CairoUniquePtr<cairo_font_options_t>(cairo_font_options_create());
```
Could we avoid the initialization here? In the case of GTK we avoid one allocation, since we always call setDefaultCairoFontOptions().

carlosgcampos commented on Source/WebKit/WebProcess/gtk/GtkSettingsManagerProxy.cpp:
```diff
@@ -62,17 +63,25 @@ void GtkSettingsManagerProxy::applySettings(GtkSettingsState&& state)
-    if (state.xftRGBA)
+    if (state.xftRGBA) {
         g_object_set(m_settings, "gtk-xft-rgba", state.xftRGBA->utf8().data(), nullptr);
+        applyFontOptions();
```
Instead fo applying font options on every change, we could add a variable `fontOptionsDidChange` set when any setting changes and then call `applyFontOptions()` once since it applies all the options.

carlosgcampos commented on Source/WebKit/WebProcess/gtk/GtkSettingsManagerProxy.cpp:
```diff
@@ -90,4 +99,70 @@ void GtkSettingsManagerProxy::applySettings(GtkSettingsState&& state)
 
+void GtkSettingsManagerProxy::applyFontOptions()
+{
+    cairo_font_options_t* fontOptions = cairo_font_options_create();
```
CairoUniquePtr

carlosgcampos reviewed (changes requested):
This is a great idea! I only have a some comments and suggestions. Thnaks!

carlosgcampos commented on Source/WebCore/platform/graphics/cairo/CairoUtilities.cpp:
```diff
@@ -49,13 +48,20 @@
+    return s_defaultCairoFontOptions.get();
+}
+
+void setDefaultCairoFontOptions(CairoUniquePtr<cairo_font_options_t> options)
```
```suggestion
void setDefaultCairoFontOptions(CairoUniquePtr<cairo_font_options_t>&& options)
```

carlosgcampos commented on Source/WebCore/platform/graphics/cairo/CairoUtilities.h:
```diff
@@ -69,9 +70,8 @@ class CairoFtFaceLocker {
-#if USE(CAIRO)
 const cairo_font_options_t* getDefaultCairoFontOptions();
-#endif
+void setDefaultCairoFontOptions(CairoUniquePtr<cairo_font_options_t>);
```
```suggestion
void setDefaultCairoFontOptions(CairoUniquePtr<cairo_font_options_t>&&);
```

carlosgcampos commented on Source/WebCore/platform/graphics/cairo/CairoUtilities.cpp:
```diff
@@ -49,13 +48,20 @@
-    static NeverDestroyed<cairo_font_options_t*> options = cairo_font_options_create();
-    return options;
+    if (!s_defaultCairoFontOptions)
+        s_defaultCairoFontOptions.reset(cairo_font_options_create());
```
hmm, I think this could be called from multiple threads with threaded rendering or offscreen canvas enabled. Setting should always happen from the main thread, so maybe can call the set for wpe from the web process initialization like gtk does to ensure the value is always set before get is called.

carlosgcampos reviewed (approved):
Excellent!

obyknovenius committed:
[GLIB] Unify code for accessing default `cairo_font_options`
https://bugs.webkit.org/show_bug.cgi?id=252057

Reviewed by Carlos Garcia Campos.

Currently, there are two different implementations of
`getDefaultCairoFontOptions()` on WPE and GTK.

WPE version simply returns a static value created by
`cairo_font_options_create()`, whereas GTK one is creating a singleton
class that stores and updates an instance variable to keep it in sync
with `GtkSettings`.

This patch makes `getDefaultCairoFontOptions()` simply return a static
value and adds `setDefaultCairoFontOptions()` function. GTK port use it
to set a new value when `GtkSettingsManagerProxy` applies new settings.

* Source/WebCore/platform/graphics/cairo/CairoUtilities.cpp:
(WebCore::getDefaultCairoFontOptions):
(WebCore::setDefaultCairoFontOptions):
* Source/WebCore/platform/graphics/cairo/CairoUtilities.h:
* Source/WebCore/platform/graphics/gtk/GdkCairoUtilities.cpp:
(WebCore::SystemFontOptions::singleton): Deleted.
(WebCore::SystemFontOptions::SystemFontOptions): Deleted.
(WebCore::SystemFontOptions::fontOptions const): Deleted.
(WebCore::SystemFontOptions::updateFontOptions): Deleted.
(WebCore::getDefaultCairoFontOptions): Deleted.
* Source/WebKit/WebProcess/gtk/GtkSettingsManagerProxy.cpp:
(WebKit::GtkSettingsManagerProxy::applySettings):
(WebKit::GtkSettingsManagerProxy::applyFontOptions):
* Source/WebKit/WebProcess/gtk/GtkSettingsManagerProxy.h:

Canonical link: https://commits.webkit.org/260301@main
